File: Extensions/fishcatcher/src/engine/aitm.js
// M8 — AiTM (adversary-in-the-middle) core. Pure: DOM-derived facts in, reasons
// out. No chrome/DOM APIs. Bundled into background.js AND probe.js by build.mjs.
// Every top-level name is aitm-prefixed for the bundler's uniqueness rule.
//
// PRIMARY detection = identity/security interaction + inferred brand + origin
// mismatch, ALL three. Proxy/keyword/hostname-substring alone never score:
// the soft corroborators (resource-graph, favicon drift) are only ever appended
// AFTER the primary fires, which structurally enforces the hard FP rule.
import { registrableDomain } from './psl.js';

// Canonical identity/security interaction kinds (Signal 1). The content script
// only ever emits values from this set.
export const AITM_INTERACTION_KINDS = ['password', 'otp', 'mfaApproval', 'deviceCode', 'passkey', 'mfaMigration', 'ssoSetup', 'qrAuth', 'helpdeskUpgrade'];

// Calibration knobs — exported so tests pin behaviour and tuning is one line.
export const AITM_PRIMARY_WEIGHT = 55; // composite strong flag → 'high' alone
export const AITM_GRAPH_WEIGHT = 15;   // Signal 2 corroborator
export const AITM_DRIFT_WEIGHT = 5;    // Signal 4 boost
const AITM_GRAPH_DOMINANCE = 0.8;      // fraction of resources under one origin
const AITM_GRAPH_MIN_RESOURCES = 5;    // don't judge tiny pages
const AITM_GRAPH_MIN_ANOMALIES = 2;    // both anomalies needed to corroborate

// One control's accessible text → interaction kind. First match wins. EN with
// cheap KA/RU tokens. Order matters: narrower prompts before the bare
// "password" fallback so "one-time password" classifies as otp, not password.
const AITM_KIND_PATTERNS = [
  ['otp', /\b(one[\s-]?time (code|password|passcode)|verification code|security code|6[\s-]?digit code|enter the code)\b|ერთჯერად|код подтвержд/i],
  ['mfaApproval', /\b((approve|deny|verify) (this )?(sign[\s-]?in|request|login)|open your authenticator|check your (phone|authenticator)|number matching|(enter|tap|select|match) the number( shown)?)\b/i],
  ['mfaMigration', /\b((migrate|re[\s-]?enroll|re[\s-]?register|update|move) your (mfa|authenticator|2fa|multi[\s-]?factor)|authenticator migration)\b/i],
  ['passkey', /\b(set ?up|create|add|register|enroll|use|sign in with) (a )?(passkey|security key|fido2? key|face ?id|touch ?id|windows hello)\b/i],
  ['ssoSetup', /\b(set ?up|configure|enable) (single sign[\s-]?on|sso|federation|saml|oidc)\b/i],
  ['qrAuth', /\bscan (this|the) qr( code)?\b.{0,40}\b(sign in|log ?in|authenticate|link|pair)\b/i],
  ['helpdeskUpgrade', /\b((security|account|mfa|password) (upgrade|migration|re[\s-]?validation|re[\s-]?verification) (required|needed)|verify your account to continue)\b/i],
  ['password', /\b(pass(word|phrase))\b|პაროლ|парол/i]
];

export function aitmKindForText(text) {
  if (!text) return null;
  for (const [kind, re] of AITM_KIND_PATTERNS) {
    if (re.test(text)) return kind;
  }
  return null;
}

// ponytail: naive IP check so registrableDomain doesn't mangle IP resource
// hosts. Upgrade to signals.js isIpAddress only if it ever moves into this bundle.
function aitmIsIp(host) {
  return /^\d{1,3}(\.\d{1,3}){3}$/.test(host) || host.includes(':') || host.startsWith('[');
}

// identityHints = { title, ogSiteName, logoAlts:[], brandTokens:[] }. The claimed
// PRIMARY identity is inferred ONLY from the page's own self-identity (title,
// og:site_name, logo alt), NOT from brandTokens (form buttons/labels): a
// legitimate SaaS's "Sign in with Google" button lives in brandTokens and would
// otherwise be read as the page claiming to BE Google, flagging every federated
// login. Conservative: empty strong hints → [].
export function aitmInferBrands(identityHints, data) {
  const h = identityHints ?? {};
  const hay = [h.title ?? '', h.ogSiteName ?? '', ...(h.logoAlts ?? [])].join(' ').toLowerCase();
  if (!hay.trim()) return [];
  const names = [];
  for (const brand of data.brands) {
    if (hay.includes(brand.name.toLowerCase()) || brand.keywords.some((k) => hay.includes(k.toLowerCase()))) {
      names.push(brand.name);
    }
  }
  return [...new Set(names)];
}

// First claimed brand whose .domains does NOT include the origin's registrable,
// else null. Allowlists are gated by the caller (runAitm) before this runs.
export function aitmClaimedMismatch(claimedBrands, registrable, data) {
  for (const name of claimedBrands) {
    const brand = data.brands.find((b) => b.name === name);
    if (brand && !brand.domains.includes(registrable)) return name;
  }
  return null;
}

// { host: count } → { registrable: count }. IP hosts pass through unchanged.
export function aitmReduceGraph(resourceHosts, data) {
  const graph = {};
  for (const host of Object.keys(resourceHosts ?? {})) {
    const key = aitmIsIp(host) ? host : registrableDomain(host, data.psl);
    graph[key] = (graph[key] ?? 0) + resourceHosts[host];
  }
  return graph;
}

// Reverse-proxy anomalies (0..2): (a) own-origin dominance; (b) every expected
// brand/IdP origin absent from the graph.
export function aitmResourceAnomalies(graph, claimedBrands, registrable, data) {
  const hosts = Object.keys(graph);
  const total = hosts.reduce((s, k) => s + graph[k], 0);
  let anomalies = 0;

  if (total >= AITM_GRAPH_MIN_RESOURCES && (graph[registrable] ?? 0) / total >= AITM_GRAPH_DOMINANCE) anomalies++;

  const expected = new Set();
  for (const name of claimedBrands) {
    const brand = data.brands.find((b) => b.name === name);
    if (brand) for (const d of brand.domains) expected.add(d);
  }
  if (data.aitmIdp) for (const d of data.aitmIdp) expected.add(d);
  if (expected.size && ![...expected].some((d) => graph[d] != null)) anomalies++;

  return anomalies;
}

// MAIN ENTRY — called by analyzer.js. ctx = { registrable, knownLegit, aitm }.
// Returns {score:0, reasons:[]} unless the full composite holds; corroborators
// are only ever appended on top of the primary.
export function runAitm(ctx, data) {
  const out = { score: 0, reasons: [] };
  const aitm = ctx.aitm;
  if (!aitm || !Array.isArray(aitm.interactions) || aitm.interactions.length === 0) return out;
  // Ultimate-trust classes never flag: known-legit, generic IdPs the attacker
  // can't own, and ZTNA/SWG/CASB wrappers that legitimately proxy other sites.
  if (ctx.knownLegit || data.aitmIdp?.has(ctx.registrable) || data.aitmMediation?.has(ctx.registrable)) return out;

  const claimed = aitmInferBrands(aitm.identityHints ?? {}, data);
  if (!claimed.length) return out;
  const mismatch = aitmClaimedMismatch(claimed, ctx.registrable, data);
  if (!mismatch) return out;

  out.score += AITM_PRIMARY_WEIGHT;
  out.reasons.push({ key: 'reasonAitmMismatch', params: [mismatch, ctx.registrable], weight: AITM_PRIMARY_WEIGHT });

  const graph = aitmReduceGraph(aitm.resourceHosts ?? {}, data);
  if (aitmResourceAnomalies(graph, claimed, ctx.registrable, data) >= AITM_GRAPH_MIN_ANOMALIES) {
    out.score += AITM_GRAPH_WEIGHT;
    out.reasons.push({ key: 'reasonAitmResourceGraph', params: [], weight: AITM_GRAPH_WEIGHT });
  }
  if (aitm.faviconCrossOrigin === true) {
    out.score += AITM_DRIFT_WEIGHT;
    out.reasons.push({ key: 'reasonAitmDrift', params: [], weight: AITM_DRIFT_WEIGHT });
  }
  return out;
}

File: Search.Kit/FishCatcher/Aitm.cs
using System.Text.RegularExpressions;
using static SearchKit.FishCatcher.JsText;

namespace SearchKit.FishCatcher;

/// M8, adversary-in-the-middle (relay) sign-in pages. Port of aitm.js.
///
/// The primary flag needs all three: an identity interaction on the page, a
/// brand the page claims to be, and an origin that isn't that brand's. A
/// proxy, a keyword or a host name alone never scores; the soft corroborators
/// (resource graph, icon drift) are only added once the primary has fired.
public static partial class Aitm
{
    public const int PrimaryWeight = 55; // the composite alone is 'high'
    public const int GraphWeight = 15;
    public const int DriftWeight = 5;
    private const double GraphDominance = 0.8; // share of resources under one origin
    private const int GraphMinResources = 5;   // don't judge tiny pages
    private const int GraphMinAnomalies = 2;   // both anomalies needed to corroborate

    public static readonly string[] InteractionKinds =
        ["password", "otp", "mfaApproval", "deviceCode", "passkey", "mfaMigration", "ssoSetup", "qrAuth", "helpdeskUpgrade"];

    private const RegexOptions I = RegexOptions.IgnoreCase | RegexOptions.CultureInvariant;

    [GeneratedRegex($@"{B}(one[{S}-]?time (code|password|passcode)|verification code|security code|6[{S}-]?digit code|enter the code){B}|ერთჯერად|код подтвержд", I)]
    private static partial Regex Otp();

    [GeneratedRegex($@"{B}((approve|deny|verify) (this )?(sign[{S}-]?in|request|login)|open your authenticator|check your (phone|authenticator)|number matching|(enter|tap|select|match) the number( shown)?){B}", I)]
    private static partial Regex MfaApproval();

    [GeneratedRegex($@"{B}((migrate|re[{S}-]?enroll|re[{S}-]?register|update|move) your (mfa|authenticator|2fa|multi[{S}-]?factor)|authenticator migration){B}", I)]
    private static partial Regex MfaMigration();

    [GeneratedRegex($@"{B}(set ?up|create|add|register|enroll|use|sign in with) (a )?(passkey|security key|fido2? key|face ?id|touch ?id|windows hello){B}", I)]
    private static partial Regex Passkey();

    [GeneratedRegex($@"{B}(set ?up|configure|enable) (single sign[{S}-]?on|sso|federation|saml|oidc){B}", I)]
    private static partial Regex SsoSetup();

    [GeneratedRegex($@"{B}scan (this|the) qr( code)?{B}{Dot}{{0,40}}{B}(sign in|log ?in|authenticate|link|pair){B}", I)]
    private static partial Regex QrAuth();

    [GeneratedRegex($@"{B}((security|account|mfa|password) (upgrade|migration|re[{S}-]?validation|re[{S}-]?verification) (required|needed)|verify your account to continue){B}", I)]
    private static partial Regex HelpdeskUpgrade();

    [GeneratedRegex($@"{B}(pass(word|phrase)){B}|პაროლ|парол", I)]
    private static partial Regex Password();

    // Narrower prompts first, so "one-time password" is otp, not password.
    private static readonly (string Kind, Func<Regex> Pattern)[] KindPatterns =
    [
        ("otp", Otp), ("mfaApproval", MfaApproval), ("mfaMigration", MfaMigration), ("passkey", Passkey),
        ("ssoSetup", SsoSetup), ("qrAuth", QrAuth), ("helpdeskUpgrade", HelpdeskUpgrade), ("password", Password),
    ];

    /// One control's accessible text → the interaction kind it asks for, or null.
    public static string? KindForText(string? text)
    {
        if (string.IsNullOrEmpty(text)) return null;
        foreach (var (kind, pattern) in KindPatterns)
            if (pattern().IsMatch(text)) return kind;
        return null;
    }

    // Naive on purpose (aitm.js): keeps the PSL from mangling IP resource hosts.
    private static bool IsIp(string host)
    {
        if (host.Contains(':') || host.StartsWith('[')) return true;
        int parts = 0;
        foreach (var range in host.AsSpan().Split('.'))
        {
            var p = host.AsSpan(range);
            if (++parts > 4 || p.Length is < 1 or > 3) return false;
            foreach (char c in p) if (c is < '0' or > '9') return false;
        }
        return parts == 4;
    }

    /// The brands the page says it is, from its own title, og:site_name and
    /// logo alt texts only. Form labels ("Sign in with Google") don't count,
    /// or every federated sign-in page would claim to be Google.
    public static List<string> InferBrands(IdentityHints hints, FishData data)
    {
        var hay = Lower(string.Join(' ', new[] { hints.Title, hints.OgSiteName }.Concat(hints.LogoAlts)));
        var names = new List<string>();
        if (IsBlank(hay)) return names;
        foreach (var brand in data.Brands)
        {
            if (hay.Contains(Lower(brand.Name), StringComparison.Ordinal) ||
                brand.Keywords.Any(k => hay.Contains(Lower(k), StringComparison.Ordinal)))
            {
                if (!names.Contains(brand.Name)) names.Add(brand.Name);
            }
        }
        return names;
    }

    private static Brand? Named(FishData data, string name) => data.Brands.FirstOrDefault(b => b.Name == name);

    /// The first claimed brand whose domains don't include this site, or null.
    public static string? ClaimedMismatch(IEnumerable<string> claimed, string registrable, FishData data)
    {
        foreach (var name in claimed)
        {
            var brand = Named(data, name);
            if (brand is not null && !brand.Domains.Contains(registrable)) return name;
        }
        return null;
    }

    /// Resource host counts folded to registrable domains (IP hosts as they are).
    public static Dictionary<string, long> ReduceGraph(IReadOnlyDictionary<string, int> resourceHosts, FishData data)
    {
        var graph = new Dictionary<string, long>(StringComparer.Ordinal);
        foreach (var (host, count) in resourceHosts)
        {
            var key = IsIp(host) ? host : data.Psl.RegistrableDomain(host);
            graph[key] = graph.GetValueOrDefault(key) + count;
        }
        return graph;
    }

    /// Reverse-proxy anomalies (0 to 2): the page's own origin serves nearly
    /// everything, and none of the claimed brand's (or any IdP's) origins appear.
    public static int ResourceAnomalies(Dictionary<string, long> graph, IEnumerable<string> claimed, string registrable, FishData data)
    {
        long total = graph.Values.Sum();
        int anomalies = 0;
        if (total >= GraphMinResources && (double)graph.GetValueOrDefault(registrable) / total >= GraphDominance) anomalies++;

        var expected = new HashSet<string>(StringComparer.Ordinal);
        foreach (var name in claimed)
            if (Named(data, name) is { } brand) expected.UnionWith(brand.Domains);
        expected.UnionWith(data.AitmIdp);
        if (expected.Count > 0 && !expected.Any(graph.ContainsKey)) anomalies++;
        return anomalies;
    }

    internal static int Run(string registrable, bool knownLegit, AitmFacts aitm, FishData data, List<Signal> reasons)
    {
        if (aitm.Interactions.Count == 0) return 0;
        // Never flagged: known-legitimate sites, identity providers an attacker
        // can't own, and ZTNA/SWG wrappers that proxy other sites for a living.
        if (knownLegit || data.AitmIdp.Contains(registrable) || data.AitmMediation.Contains(registrable)) return 0;

        var claimed = InferBrands(aitm.IdentityHints, data);
        if (claimed.Count == 0) return 0;
        var mismatch = ClaimedMismatch(claimed, registrable, data);
        if (mismatch is null) return 0;

        int score = PrimaryWeight;
        reasons.Add(new Signal("reasonAitmMismatch", PrimaryWeight, mismatch, registrable));

        var graph = ReduceGraph(aitm.ResourceHosts, data);
        if (ResourceAnomalies(graph, claimed, registrable, data) >= GraphMinAnomalies)
        {
            score += GraphWeight;
            reasons.Add(new Signal("reasonAitmResourceGraph", GraphWeight));
        }
        if (aitm.FaviconCrossOrigin)
        {
            score += DriftWeight;
            reasons.Add(new Signal("reasonAitmDrift", DriftWeight));
        }
        return score;
    }
}
